docs/SETTINGS.md
# App Settings & Configuration

![command_1](https://github.com/imbateam-gg/titan-reactor/assets/586716/9709ff59-8943-402f-a5a5-f57d64a5ed81)

## Audio Settings
- **Global Volume**: Adjusts overall volume, range from 0 to 1.
- **Music Volume**: Sets music volume, range from 0 to 1.
- **Sound Effects**: Controls sound effect volume, range from 0 to 1.
- **Play Intro Sounds**: Toggles introductory sounds (wraiths).

## Graphics Settings
- **Pixel Ratio**: Sets the resolution scale, ranges from 0.5 to max setting. Set this to max for best quality.
- **Use HD2**: Controls usage of half resolution textures, options include "As Mipmap", "Never. Only HD.", and "Exclusively".
- **Preload Assets**:  Preloads graphical assets for smoother in game loading.
- **Cursor Size**: Sets cursor size, range from 0.5 to 4.

## Minimap Settings
- **Enabled**: Controls minimap visibility.
- **Interactive**: Enables interactive features on the minimap.
- **Soft Edges**: Applies soft edges to the minimap.
- **Scale**: Sets the size of the minimap, ranges from 1 to 10.
- **Position**: Sets minimap position, supports small steps.
- **Rotation**: Controls minimap rotation, supports small steps.
- **Opacity**: Sets minimap opacity, ranges from 0 to 1.

## Input Settings
- **Scene Controller**: Sets default scene controller from available plugins. ie The plugin that controls the view and view inputs.
- **Damping Factor**: Controls camera movement fluidity, range from 0.01 to 0.1.
- **Movement Speed**: Sets camera speed.
- **Rotate Speed**: Controls camera rotation speed.
- **Camera Shake Strength**: Sets camera shake intensity, range from 0 to 1.
- **Zoom Levels**: Specifies zoom levels for the camera.
- **Unit Selection**: Enables unit selection, hidden by default. Used by plugins.
- **Cursor Visible**: Toggles cursor visibility, hidden by default. Used by plugins.

## Utilities
- **Debug Mode**: Enables debug features.
- **Sanity Check Replay Commands**: Checks and rewrites command buffer overflows.
- **Detect Melee Observers**: Automatically removes melee observers from the player list.
- **Melee Observers Threshold**: Sets command threshold for observer detection, range from 1000 to 50000.
- **Alert Desynced**: Enables desync detection before the game starts.
- **Desynced Threshold**: Sets idle units threshold for desync alerts, range from 10 to 100.

## Postprocessing Settings
- **Anisotropy**: Sets texture anisotropy level, range based on max settings.
- **Antialias**: Controls anti-aliasing level, range based on max settings.
- **Bloom**: Sets bloom intensity, ranges from 0 to 10, hidden in production.
- **Brightness**: Controls brightness, range from -0.5 to 0.5.
- **Contrast**: Sets contrast, range from -0.5 to 0.5.
- **Fog of War**: Sets fog of war opacity, ranges from 0 to 1.

## Postprocessing 3D Settings
- **Tone Mapping**: Sets tone mapping exposure, ranges from 1 to 2, hidden in production.
- **Depth Focal Length**: Sets depth focal length, range from 1 to 20, hidden in production.
- **Depth Focal Range**: Sets depth focal range, range from 1 to 20, hidden in production.
- **Depth Bokeh Scale**: Controls bokeh scale, range from 1 to 5, hidden in production.
- **Depth Blur Quality**: Sets depth blur quality with preset options.
- **Environment Map**: Controls environment mapping, range from 0 to 2, hidden in production.
- **Sunlight Position**: Sets the position of sunlight, hidden in production.
- **Sunlight Intensity**: Controls sunlight intensity, range from 0 to 20, hidden in production.
- **Sunlight Color**: Sets the color of sunlight, hidden in production.
- **Shadow Quality**: Sets shadow quality, range from 0 to 8, hidden in production.


## Session Settings - Used by plugins & Macros
- **Type**: Determines the viewer's mode, options are Replay, Live, or Map.
- **Sandbox**: Enables sandbox mode, hidden by default.
- **Audio Listener Distance**: Sets the distance for 3d positional audio, ranges between 0 to 1. Mainly used by view controller plugins.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [unreleased]

## Added

- Plugin API documentation in https://imbateam-gg.github.io/titan-reactor/

## [0.6.9] - 2023-09-10

## Added

-   Two more macros oriented around following / unfollowing units
    -   Clicking the minimap unfollows the units
    -   F key will follow selected units or unfollow all units
-   Macros now have a description field making it easier to understand and explain what a macro does.
-   Added font size options to official plugins that were missing them

## Fixed

-   Some plugin changes weren't being reflected in the UI

## Changed

-   API namespace finalized to @titan-reactor-runtime/ui and @titan-reactor-runtime/native
-   Plugin file structure has changed so that typescript types can be used

## [0.6.8] - 2022-12-21

## Added

-   Additional messaging when updating / installing plugins
-   Type definitons for the UI api
-   customEvents object replacing custom callback methods
-   Spinny loader for better loading feedback
-   Code architecture documentation in WIKI
-   Load & display remote plugin README files

## Removed

-   Clear asset cache menu item
-   Plugin callbacks no longer have context value attached

## Changed

-   Last plugin tab selected will persist (Info or Configure)

## [0.6.7] - 2022-10-28

## Added

-   Macro groups

## Fixed

-   Bunker overlay offset (marine fire)
-   Fix issue where the app would "crash" on loading due to some changes in latest electron api (memory cage in v21)

## Changed

-   Plugin permissions changed from under `config` to `root` of package.json

## [0.6.6] - 2022-10-22

## Added

-   Exposed default enabled/disabled switch for macros
-   Game & sandbox modes for macro condition testing
-   Hover cursor over units, like in the game
-   Selection circles around 3d models - initial version
-   New apis for macros getHoveredUnits, and getFollowedUnitsPosition
-   Blur transition between render modes

## Changed

-   Improved default macros
-   Unlimited folders for macros

## Fixed

-   Several small optimizations
-   Audio no longer has silences (regression in 0.6.0)
-   Small visual glitch with creep colony construction and other images that shared atlases
-   Big improvement (~5s) on initial loading screen

## [0.6.5] - 2022-10-06

## Added

-   Macros can execute other macros
-   Active Preview option for modifying macros
-   IScriptah now semi functional again (Tools -> Iscriptah)

## Fixed

-   Several memory leaks due to untested Janitor implementation
-   Squashed some bugs around macros and settings
-   Units are still selectable when the game is paused

## Changed

-   Upgraded to Electron 20, TypeScript 4.8.4, Vite build system and others.
-   Initial macros and plugins setup is automatic and transparent

## [0.6.4] - 2022-09-19

## Added

-   New post processing and post processing 3d options
-   New minimap & minimap options
-   Macros now show cute little keyboard preview ^^
-   New world event system
-   New macro trigger - mouse click
-   New graphics settings (HD2 options)
-   Cursor is now animated and has cursor size option

## Changed

-   Lots of internal refactoring in preparation for sandbox mode and 3d models
-   Moved camera damping/rotation speed options to global settings
-   Additional error handling
-   Camera starts at a player location
-   Map viewer now uses OpenBW so all images/units/sprites should work
-   Fixed a bug in bw-casclib, allowing mac/linux builds to work now.

## Fixed

-   Default macros automatically install on first run
-   App no longer needs a full reload on initial plugin install
-   Bullet tracking is now more precise

## [0.6.0] - 2022-08-20

## Added

-   New home screen and map/replay loading screens
-   Macro hotkey system with "programmable" triggers and action sequences
-   New global options: minimap size, asset preload, default scene controller, anisotropy
-   Terrain now has mirrored edge tiles
-   Plugins API v2

## Changed

-   Global Settings is no longer a plugin and now built into the app
-   Menus have been re-arranged. Plugin Settings is now called Command Center.
-   More optimizations to reduce GC usage, caused minor bugs in audio which will be addressed
-   HD is now the default (rather than HD2)
-   Flying units are further affected by terrain height
-   Deprecated plugins will not show in online tab
-   Shadows will be placed in default locations unless in "3d" mode

### Fixed

-   Regression in floating sprites is fixed
-   Regression creep edges fixed
-   Sprite overlay edges bug fixed
-   Bullet tracking now works on floating buildings
-   Regression on HP bar on resources now fixed
-   Follow unit now unfollows on minimap click
-   Previously on 0.5.3 release:
    -   Tank mode siege tank turrets will now look better when camera rotated
    -   When the camera target is above the horizon units will not rotate oddly
    -   Disposing of the previous replay should execute before any additional plugin calls
    -   UMS games that use set alliance now work (thank you Heinerman!)
    -   Maps that use disabled doodads now work (thank you Heinerman!)
    -   Observer chat won't break replays anymore

### Removed

-   HD2 is no longer an option
-   Plugin API:
    -   Deprecated getFPS.
    -   Deprecated maxFrame and time fields from the useFrame() hook, use useReplay() and getFriendlyTime() instead
-   Keyboard shortcuts are no longer part of plugins and are now configured with the hotkey macro system.

## [0.5.3] - 2022-07-14

### Added

-   Warp flash animation on protoss buildings
-   skipForward, skipBackward, speedUp, speedDown, togglePause plugin apis now return appropriate values
-   speed changes now goes by 0.25 increments when >= 1 and <= 2

### Changed


src/main/settings.js
import { EventEmitter } from "events";
import { promises as fsPromises } from "fs";
import {
  RenderMode,
  ShadowLevel,
  ProducerWindowPosition,
  GameAspect,
} from "common/settings";
import isDev from "electron-is-dev";

import {
  findMapsPath,
  findReplaysPath,
  findStarcraftPath,
} from "./starcraft/findInstallPath";
import fileExists from "./utils/fileExists";
import path from "path";
const supportedLanguages = ["en-US", "es-ES", "ko-KR", "pl-PL", "ru-RU"];

const VERSION = 1;

const getEnvLocale = (env = process.env) => {
  return env.LC_ALL || env.LC_MESSAGES || env.LANG || env.LANGUAGE;
};

const shallowDiff = (a, b) => {
  const x = {};
  for (let n of Object.getOwnPropertyNames(a)) {
    if (a[n] !== b[n]) {
      x[n] = b[n];
    }
  }
  return x;
};

export class Settings extends EventEmitter {
  constructor(filepath) {
    super();
    this._settings = {};
    this._filepath = filepath;
    this._initialized = false;
  }

  async init(webGLCapabilities) {
    try {
      // await fsPromises.unlink(this._filepath);

      this._settings = JSON.parse(
        await fsPromises.readFile(this._filepath, { encoding: "utf8" })
      );
    } catch (err) {
      try {
        await fsPromises.unlink(this._filepath);
      } catch (err) {
      } finally {
        await this.save(await this.createDefaults(webGLCapabilities));
      }
    }
    this._initialized = true;
  }

  async get() {
    if (!this._initialized) {
      throw new Error("settings not initialized");
    }
    const errors = [];
    const files = [
      "starcraftPath",
      "mapsPath",
      "replaysPath",
      "communityModelsPath",
    ];

    for (let file of files) {
      if (!(await fileExists(this._settings[file]))) {
        errors.push(file);
      }
    }

    const dataFolders = [
      "anim",
      "arr",
      "cursor",
      "game",
      "music",
      "portrait",
      "scripts",
      "sound",
      "TileSet",
      "unit",
    ];
    if (await fileExists(this._settings["starcraftPath"])) {
      for (let folder of dataFolders) {
        if (
          !(await fileExists(
            path.join(this._settings["starcraftPath"], folder)
          ))
        ) {
          if (!errors.includes("starcraftPath")) {
            errors.push("starcraftPath");
          }
        }
      }
    }

    const localLanguage = supportedLanguages.includes(getEnvLocale())
      ? getEnvLocale()
      : "en-US";
    this._settings.language = supportedLanguages.includes(
      this._settings.language
    )
      ? this._settings.language
      : localLanguage;

    const phrases = await import(`common/phrases/${this._settings.language}`);

    return { data: this._settings, errors, isDev: true, phrases, diff: {} };
  }

  async load() {
    if (!this._initialized) {
      throw new Error("settings not initialized");
    }
    const contents = await fsPromises.readFile(this._filepath, {
      encoding: "utf8",
    });
    const newData = JSON.parse(contents);
    this._settings = newData;
    this._emitChanged();
  }

  async save(settings) {
    if (settings.errors) {
      delete settings.errors;
    }

    const diff = shallowDiff(this._settings, settings);
    this._settings = Object.assign({}, this._settings, settings);
    await fsPromises.writeFile(this._filepath, JSON.stringify(this._settings), {
      encoding: "utf8",
    });
    this._emitChanged(diff);
  }

  async _emitChanged(diff = {}) {
    this.emit("change", { ...(await this.get()), diff });
  }

  async createDefaults(webGLCapabilities) {
    return {
      version: VERSION,
      renderMode: RenderMode.SD,
      maxAutoReplaySpeed: 1.2,
      alwaysHideReplayControls: true,
      startPaused: false,
      language: supportedLanguages.includes(getEnvLocale())
        ? getEnvLocale()
        : "en-US",
      starcraftPath: await findStarcraftPath(),
      mapsPath: await findMapsPath(),
      replaysPath: await findReplaysPath(),
      communityModelsPath: "",
      observerLink: "",
      musicVolume: 0.01,
      soundVolume: 1,
      antialias: false,
      anisotropy: 1,
      pixelRatio: 1,
      maxAnisotropy: webGLCapabilities.anisotropy,
      gamma: 1.2,
      shadows: ShadowLevel.High,
      shadowTextureSize: 0.5,
      keyPanSpeed: 0.5,
      twitch: "",
      countdownTimer: true,
      cameraStyle2dOrtho: true,
      cameraStyle3dOrtho: false,
      producerWindowPosition: ProducerWindowPosition.None,
      producerDockSize: 300,
      gameAspect: GameAspect.Fit,
      fullscreen: true,
      enablePlayerScores: true,
      esportsHud: true,
      showTooltips: true,
      cameraShake: 1,
      mapsRss: "",
      replaysRss: "",
      useCustomColors: false,
      usePlayerColorUnitSelection: true,
      player1Color: "#ef4444",
      player2Color: "#3b82f6",

      hudFontSize: "sm",
      esportsHudSize: "lg",
      minimapRatio: 25,
      replayAndUnitDetailSize: "24vw",
      showWorkerCount: true,
      showArmyCount: true,
    };
  }
}
